File: muxdb-js/src/balancer.ts
export interface BalancingAction {
  actionType: 'L1_LEASE' | 'L2_SPLIT' | 'L3_REPLICA';
  shardId: string;
  details: Record<string, any>;
}

export class Balancer {
  private lastBalanceTime: number = 0;

  constructor(
    public readonly cooldownSeconds: number = 10,
    public readonly qpsSplitThreshold: number = 100,
    public readonly qpsReplicaThreshold: number = 50,
    public readonly writeSuppressionThreshold: number = 80
  ) {}

  public evaluate(metrics: Record<string, { read_qps: number; write_qps: number }>): BalancingAction[] {
    /** Evaluate metrics and trigger range splits, replica scales, or lease transfers. */
    const now = Date.now() / 1000;
    if (now - this.lastBalanceTime < this.cooldownSeconds) {
      return [];
    }

    // Write load suppression check
    const totalWriteQps = Object.values(metrics).reduce((sum, m) => sum + (m.write_qps || 0), 0);
    if (totalWriteQps >= this.writeSuppressionThreshold) {
      return [];
    }

    const actions: BalancingAction[] = [];

    for (const [shardId, data] of Object.entries(metrics)) {
      const readQps = data.read_qps || 0;
      const writeQps = data.write_qps || 0;
      const totalQps = readQps + writeQps;

      // L2 Split check
      if (totalQps >= this.qpsSplitThreshold) {
        actions.push({
          actionType: 'L2_SPLIT',
          shardId,
          details: { qps: totalQps, splitThreshold: this.qpsSplitThreshold }
        });
        continue;
      }

      // L3 Replica scaling check
      if (readQps >= this.qpsReplicaThreshold) {
        actions.push({
          actionType: 'L3_REPLICA',
          shardId,
          details: { readQps, replicaThreshold: this.qpsReplicaThreshold }
        });
        continue;
      }

      // L1 Lease Transfer check
      if (writeQps > 10 && readQps > 20) {
        actions.push({
          actionType: 'L1_LEASE',
          shardId,
          details: { readQps, writeQps }
        });
      }
    }

    if (actions.length > 0) {
      this.lastBalanceTime = now;
    }

    return actions;
  }
}

File: muxdb-py/muxdb/balancer.py
from __future__ import annotations

import time
from typing import Any, Dict, List

class BalancingAction:
    """Represents a load balancing action recommended by the Balancer."""

    def __init__(self, action_type: str, shard_id: str, details: Dict[str, Any]) -> None:
        self.action_type = action_type  # "L1_LEASE", "L2_SPLIT", "L3_REPLICA"
        self.shard_id = shard_id
        self.details = details

    def __repr__(self) -> str:
        return f"BalancingAction(type={self.action_type}, shard={self.shard_id}, details={self.details})"


class Balancer:
    """Evaluates telemetry metrics to trigger dynamic range splits, replica scales, or lease transfers."""

    def __init__(
        self,
        cooldown_seconds: float = 10.0,
        qps_split_threshold: float = 100.0,
        qps_replica_threshold: float = 50.0,
        write_suppression_threshold: float = 80.0,
    ) -> None:
        self.cooldown_seconds = cooldown_seconds
        self.qps_split_threshold = qps_split_threshold
        self.qps_replica_threshold = qps_replica_threshold
        self.write_suppression_threshold = write_suppression_threshold
        self.last_balance_time = 0.0

    def evaluate(self, metrics: Dict[str, Dict[str, float]]) -> List[BalancingAction]:
        """Analyze shard metrics and generate balancing actions.

        Applies suppression rule under high write-throughput or cooldown state.
        """
        now = time.time()
        if now - self.last_balance_time < self.cooldown_seconds:
            return []

        # Write load suppression check
        total_write_qps = sum(m.get("write_qps", 0.0) for m in metrics.values())
        if total_write_qps >= self.write_suppression_threshold:
            return []

        actions: List[BalancingAction] = []

        for shard_id, data in metrics.items():
            read_qps = data.get("read_qps", 0.0)
            write_qps = data.get("write_qps", 0.0)
            total_qps = read_qps + write_qps

            # Check L2 Split
            if total_qps >= self.qps_split_threshold:
                actions.append(
                    BalancingAction(
                        action_type="L2_SPLIT",
                        shard_id=shard_id,
                        details={"qps": total_qps, "split_threshold": self.qps_split_threshold},
                    )
                )
                continue

            # Check L3 Replica Scaling
            if read_qps >= self.qps_replica_threshold:
                actions.append(
                    BalancingAction(
                        action_type="L3_REPLICA",
                        shard_id=shard_id,
                        details={"read_qps": read_qps, "replica_threshold": self.qps_replica_threshold},
                    )
                )
                continue

            # Check L1 Lease Transfer (read lease offloaded if write-heavy)
            if write_qps > 10.0 and read_qps > 20.0:
                actions.append(
                    BalancingAction(
                        action_type="L1_LEASE",
                        shard_id=shard_id,
                        details={"read_qps": read_qps, "write_qps": write_qps},
                    )
                )

        if actions:
            self.last_balance_time = now

        return actions
